Tutorial for Ground Consistency Layer (#911)
ros-planning/navigation.ros.org: tutorials/docs/navigation2_with_ground_consistency_layer.rst

==========================================

> NAV2 Ground Consistency Demo

Reviewer: Perhaps change to `Ground Terrain Segmentation using 3D Lidar`? 

ALso the `====` has to match the char length of the title to render properly

==========================================

> A tutorial demonstrating terrain-aware navigation using 3D lidar ground segmentation with Nav2's `ground consistency costmap layer <https://github.com/dfki-ric/nav2_ground_consistency_costmap_plugin>`_. Learn how to classify terrain into traversable ground and obstacles, then use that classification to build smarter costmaps for safer navigation.

Reviewer: ```suggestion
This tutorial demonstrates terrain-aware navigation using 3D lidar ground segmentation with the `ground consistency costmap layer <https://github.com/dfki-ric/nav2_ground_consistency_costmap_plugin>`_. This allows users to classify terrain into traversable ground and obstacles for intelligent outdoor navigation in non-planar environments for smarter costmaps for safer navigation.
```

.. image:: images/Navigation2_with_Ground_Consistency/ground_consistency_layer.gif
   :alt: Ground Consistency Layer Demo

Overview
========

This tutorial demonstrates a complete workflow for terrain-aware navigation:

1. **Gazebo Simulation** - A Husky robot with VLP-16 lidar in realistic terrain
2. **Ground Segmentation** - Classifies every lidar point as ground or obstacle
3. **Height-Based Filtering** - Distinguishes actual blocking obstacles from overhead structures
4. **Nav2 Integration** - Uses segmentation to build cost maps that understand terrain
5. **Smart Navigation** - Robot avoids real obstacles while safely navigating complex terrain

Requirements
============

> It is assumed you have ROS 2 Jazzy installed. To install all required dependencies and clone necessary repositories, run:

Reviewer: See other tutorials & link to the getting started guide if they don't have Nav2 setup/installed yet. 

Then: `Then, install the required depedencies:`

.. code-block:: bash

   cd ~/path_to_my_custom_workspace
   mkdir -p src && cd src
   git clone https://github.com/ros-navigation/navigation2_tutorials.git && cd navigation2_tutorials
   git checkout -t remotes/origin/jazzy

.. code-block:: bash

   cd nav2_ground_consistency_demo

>    bash install_dependencies.bash ~/path_to_my_custom_workspace

Reviewer: This needs to be updated/fixed. We should be using `rosdep` for any dependencies that are installed in `apt`. If we need to build other things from source, use `.repos` files.

Overall this installation should basically read:

* Install ROS 2 / Nav2 --> if not link to the getting started pages (see other tutorials)
* Then, install the source code for the tutorial used --> clone & rosdep & repos install

This script will:

- Install required ROS 2 packages (Gazebo, visualization, navigation)
- Clone all necessary repositories (KISS-ICP, ground segmentation, etc.)
- Provide next steps for building

Ground Segmentation Overview
============================

**What is Ground Segmentation?**

Ground segmentation is the first step in terrain-aware navigation. It takes raw 3D lidar point cloud data and classifies every point into two categories:

- **Ground Points**: Points that lie on the terrain surface (traversable)

> - **Non-Ground Points**: Points above the terrain surface (potential obstacles)

Reviewer: ```suggestion
Ground segmentation is the first step in terrain-aware navigation. It takes sensor data and classifies its readings into at least two categories: ground and non-ground. This can be from raw 3D lidar point clouds as used in this tutorial, but also ground segmentation may be the result of RGBD cameras, monocular vision, or other source using geometric of AI based technologies.
```

> This is done by analyzing the geometry of the point cloud - points that are approximately at the same height and form a plane are ground, while points sticking up above this plane are classified as obstacles. The algorithm doesn't require a pre-built map; it works in real-time as the robot moves.

Reviewer: ```suggestion
This demonstration analyzes the geometry of a 3D lidar point cloud using the XYZ library (provide a link TODO). Points that are approximately at the same height and form a plane are ground, while points sticking up above this plane are classified as obstacles. The algorithm doesn't require a pre-built map; it works in real-time as the robot moves.
```

I'd add a paragraph after this briefly talking about your library / method. Keep it high level, but some detail may be nice :-)

**Why is this important?** Without ground segmentation, a navigation system can't tell the difference between actual obstacles and normal terrain variations (like slopes or small rocks). Ground consistency uses this classification to make smarter navigation decisions.

Ground Consistency Overview
===========================

What is Ground Consistency?
---------------------------

Traditional occupancy grids treat all obstacles equally. A 10cm tall rock and a 2m tall wall both show up as "occupied." This causes problems:

- **On slopes**: Real traversable ground is marked as obstacles
- **Under structures**: Bridges and overpasses block navigation unnecessarily
- **In clutter**: Small harmless debris at ground level makes navigation overly cautious

> **Ground consistency** solves this by adding a height dimension to costmaps. It answers: *"Is this obstacle actually blocking my robot, or is it something I can navigate over/under?"*

Reviewer: Can this not just go under the Ground Segmentation Overview section?

How It Works

> ------------

Reviewer: This section is way too TMI. Please distill this into 2 short paragraphs instead. You don't need to describe the entire algorithm, just generally how to think about it and how this works.

   # Check ROS 2 packages
   ros2 pkg list | grep -E "ground_segmentation|kiss_icp|nav2_ground_consistency"

You should see:

- ``ground_segmentation_ros2``
- ``kiss_icp``
- ``nav2_ground_consistency_costmap_plugin``
- ``nav2_ground_consistency_demo``

Step 1: Launch Gazebo Simulation
---------------------------------

In **Terminal 1**, start the Gazebo simulation with the Husky robot:

Note: Gazebo will take some time to download the baylands terrain and husky model

.. code-block:: bash

   source install/setup.bash
   ros2 launch nav2_ground_consistency_demo start.launch.py

You should see:

- Gazebo opens with Baylands terrain
- Husky robot appears on the ground
- ROS topics ``/husky/scan/points``, ``/husky/imu`` being published

.. image:: images/Navigation2_with_Ground_Consistency/husky_in_baylands.png
   :alt: Husky in Baylands

Verify the lidar is working:

.. code-block:: bash

   ros2 topic hz /husky/scan/points

You should see ~5-10 Hz updates based on your computer specifications. You will need to set your rviz 'Frame Rate' to this value to avoid issues with rviz visualizations.

.. image:: images/Navigation2_with_Ground_Consistency/rviz_frame_settings.png

>    :alt: RViz Frame Settings

Reviewer: Remove this please, TMI

Step 2: Launch Navigation Stack
--------------------------------

In **Terminal 2**, launch the stack with the ground consistency layer (this includes ground segmentation):

On some systems, the ``ground_segmentation_ros2_node`` may fail to start with: ``error while loading shared libraries: libjawt.so: cannot open shared object file``

Fix this error by using the following:

.. code-block:: bash

   export JAVA_HOME=$(dirname $(dirname $(readlink -f $(which javac))))
   echo "$JAVA_HOME/lib" | sudo tee /etc/ld.so.conf.d/java.conf
   echo "$JAVA_HOME/lib/server" | sudo tee -a /etc/ld.so.conf.d/java.conf

>    sudo ldconfig

Reviewer: This... shouldn't be necessary. Why?

.. code-block:: bash

   source install/setup.bash
   ros2 launch nav2_ground_consistency_demo full_stack.launch.py

.. code-block:: bash

   ros2 lifecycle set /controller_server configure

>    ros2 lifecycle set /controller_server activate

Reviewer: These should change if/when using Nav2. You also wouldn't need to manually enable the controller server

- **Controller Server** - Path tracking and obstacle avoidance
- **RViz2** - Visualization (auto-starts)
- **Lifecycle Transitions** - Automatic state management

You should see RViz open with:

- Robot footprint (yellow rectangle)
- Lidar pointcloud (cyan points)
- Local costmap
- Ground consistency layer visualization

**Verify that everything is working** in **Terminal 3**:

.. code-block:: bash

   # Check KISS-ICP odometry
   ros2 topic hz /tf

   # Check ground segmentation output
   ros2 topic hz /ground_segmentation/ground_points
   ros2 topic hz /ground_segmentation/obstacle_points

You should see:

- ``/tf`` publishing at ~5-10 Hz (odometry transforms)
- Ground points publishing at ~5-10 Hz (from ground segmentation)
- Obstacle points publishing at ~5-10 Hz (from ground segmentation)

If any topic shows "0 Hz", something is not working - check the logs in the launch terminal for errors.

Step 3: Move the Robot
----------------------

To move the robot around and see ground consistency in action:

In **Gazebo**:

1. Click the menu (⋮) in the top-right corner
2. Search for "Teleop" and select the teleop widget
3. Change the topic field to ``/model/husky/cmd_vel``

> 4. Use the sliders or buttons to move the robot forward/backward and rotate

Reviewer: Use Nav2 via the navigat to pose in rviz instead, this demo is to show how to use this layer with nav2, so you should use nav2 😉

Watch the costmap in RViz update in real-time as the robot moves:

- **On flat ground**: Costmap shows mostly empty (safe)
- **On slopes**: Ground segmentation classifies the slope as traversable (not blocked)
- **Over small obstacles**: Debris below robot height doesn't appear as lethal obstacles
- **Under overpasses**: If your world has them, the robot can navigate underneath

> 

Reviewer: Here's a good place to add a youtuve video or two embedded

Author: I am preparing a video, will update you on it

Reviewer: OK!

References
==========

- `NAV2 Documentation <https://docs.nav2.org/>`_
- `Ground Consistency Layer Plugin <https://github.com/dfki-ric/nav2_ground_consistency_costmap_plugin>`_
- `Ground Segmentation ROS2 <https://github.com/dfki-ric/ground_segmentation_ros2>`_
- `KISS-ICP Odometry <https://github.com/prbonn/kiss-icp>`_
- `Gazebo Simulator <https://gazebosim.org/>`_
- `Clearpath Husky Robot <https://clearpathrobotics.com/husky>`_
- `ROS 2 Jazzy Documentation <https://docs.ros.org/en/jazzy/>`_

KISS-ICP Odometry
-----------------

KISS-ICP is used as an odometry solution in the demo.

.. code-block:: bibtex

   @article{vizzo2023ral,
   author    = {Vizzo, Ignacio and Guadagnino, Tiziano and Mersch, Benedikt and Wiesmann, Louis and Behley, Jens and Stachniss, Cyrill},
   title     = {{KISS-ICP: In Defense of Point-to-Point ICP -- Simple, Accurate, and Robust Registration If Done the Right Way}},
   journal   = {IEEE Robotics and Automation Letters (RA-L)},
   pages     = {1029--1036},
   doi       = {10.1109/LRA.2023.3236571},
   volume    = {8},
   number    = {2},
   year      = {2023},
   codeurl   = {https://github.com/PRBonn/kiss-icp},

>    }

Reviewer: Remove please, this isn't an academic paper

Husky Robot Model
-----------------

> This demo uses a local copy of the COSTAR_HUSKY_SENSOR_CONFIG_1 model from Gazebo Fuel. The local copy has been modified to enable IMU orientation output (``enable_orientation=1``) for proper quaternion-based orientation estimates, which is required for ground segmentation with IMU integration. A DiffDrive plugin was included in the robot model which executes motion commands sent to the robot.

Reviewer: Huh?

**Original Model Citation:**

.. code-block:: bibtex

   @online{GazeboFuel-OpenRobotics-COSTAR_HUSKY_SENSOR_CONFIG_1,
       title={COSTAR_HUSKY_SENSOR_CONFIG_1},
       organization={Open Robotics},
       date={2023},
       month={September},
       day={3},
       author={OpenRobotics},
       url={https://fuel.gazebosim.org/1.0/OpenRobotics/models/COSTAR_HUSKY_SENSOR_CONFIG_1},

>    }

Reviewer: Remove this

   :alt: Ground Consistency Layer Demo

*Ground Consistency Layer in action: Green points are ground, magenta points are obstacles overlaid on the costmap*

Requirements
============

Install ROS 2, Nav2, and Ground Segmentation
--------------------------------------------

This tutorial requires ROS 2 Jazzy and Nav2. Follow the official setup guides:

- **ROS 2 Jazzy Installation**: https://docs.ros.org/en/jazzy/Installation.html
- **Nav2 Setup Guide**: https://docs.nav2.org/getting_started/index.html#installation
- **Ground Segmentation**: There are many well-established ground segmentation methods; in this tutorial, we use the `ground_segmentation_ros2 <https://github.com/dfki-ric/ground_segmentation_ros2>`_ package developed by `DFKI Robotics Innovation Center <https://robotik.dfki-bremen.de/en/startpage>`_. You will find the installation instructions `here <https://github.com/dfki-ric/ground_segmentation_ros2#prerequisite>`_

Installation Steps
==================

It is assumed that ROS 2, Nav2, and a ground segmentation package are built locally. In case you want to skip to the practical example, you can follow the

> instructions in the `Practical Example`_ section below, which will set up everything for you.

Reviewer: This is the installation step, we should be explaining how to install the code needed for this tutorial. This shouldn't be punted to another section at the end please. Move the "Install Required Tools" through "Setup Simulation Environment"'s build instructions so that we have everything downloaded, built, and ready for users to follow along.

Then move the demo stuff before the conclusion.

This should be linear, step by step. It is not assumed that the ground segmentation package is already built, show them with the full installation instructions in the Installation Instructions section.

Reviewer: Also, add a "If you don't have Nav2 already installed, check out our Getting Started page before moving forward." (with a hyperlink on the Getting Started page)

Next, you will also need to compile the `nav2_ground_consistency_costmap_plugin <https://github.com/dfki-ric/nav2_ground_consistency_costmap_plugin>`_ package. To do so, clone the repository to your ROS 2 workspace and build the package:

> Note: currently only ROS 2 Jazzy (main branch) is supported, but branches for humble and rolling will be added soon.

Reviewer: Remove, this dates things and if its going to be soon, then its fine to bypass

Author: Removed

.. code-block:: bash

   cd <your workspace path>/src
   git clone https://github.com/dfki-ric/nav2_ground_consistency_costmap_plugin.git
   cd <your workspace path>
   colcon build --symlink-install # on your workspace path

Ground Segmentation Overview
============================

Ground segmentation in point cloud data is the

> process of separating ground points from obstacle points.

Reviewer: ```suggestion
process of separating ground points from obstacle points when not operating on trivial, planar terrain.
```

This task is fundamental for perception in mobile robotics, where safety and reliable operation depend on the
precise detection of obstacles and navigable surfaces. A single LiDAR scan
can contain over a million points, capturing both ground and
obstacle surfaces within the region of interest. However,
not all points are equally relevant for downstream perception
tasks. For example, object and obstacle detection algorithms often
misclassify ground points as obstacles, leading to false positives that can compromise safety and increase computational
load. Accurate ground segmentation is essential for reliable perception, traversability analysis, navigation, and map
generation in autonomous systems. By effectively distinguishing between ground and obstacle points, we can create more

> accurate costmaps that allow the robot to navigate safely and efficiently in complex environments.

Reviewer: ```suggestion
tasks. Accurate ground segmentation is essential for reliable perception, traversability analysis, navigation, and map generation in autonomous systems. By effectively distinguishing between ground and obstacle points, we can create more accurate costmaps that allow the robot to navigate safely and efficiently in complex environments.
```

**Benefits of Ground Segmentation:**

> - **Reduced false positives**: Eliminates misclassified ground points that would otherwise block navigation

Reviewer: ```suggestion
```

I think this is an odd point to make given what we're doing in this integration

- **Improved computational efficiency**: Filters irrelevant points before processing, reducing memory and CPU usage
- **Enhanced traversability analysis**: Accurately identifies traversable terrain for safe navigation planning
- **Better obstacle detection**: Focuses processing on actual obstacles rather than terrain variations
- **Robust slope handling**: Distinguishes between navigable slopes and actual blocking obstacles
- **Underground/tunnel navigation**: Enables navigation in challenging environments where traditional costmaps fail

> 

Reviewer: ```suggestion
- **Enhanced traversability analysis**: Accurately identifies traversable terrain for safe navigation planning
- **Better obstacle detection**: Focuses processing on actual obstacles rather than terrain variations
- **Robust slope handling**: Distinguishes between navigable slopes and actual blocking obstacles
- **Improved computational efficiency**: Filters irrelevant points before processing, reducing memory and CPU usage
- **Underground/tunnel navigation**: Enables navigation in challenging environments where traditional costmaps fail

```

Lets not bury the lead of the main points: navigable terrain identification

.. image:: images/Navigation2_with_Ground_Consistency/raw_points.png
   :alt: Raw Input Point Cloud

*Raw Input Point Cloud*

.. image:: images/Navigation2_with_Ground_Consistency/segmented_points.png
   :alt: Segmented Ground and Obstacle Points

*Segmented Ground (Green) and Obstacle Points (Magenta) based on* `GSeg3D <https://github.com/dfki-ric/ground_segmentation_ros2>`_

While ground segmentation is traditionally used to filter out ground points from costmaps, we can instead use ground points in our decision-making process to implement an
evidence accumulation and height-based classification system. This approach creates more accurate and stable costmaps in outdoor challenging environments.

Ground Consistency Layer
========================

Ground Consistency is a costmap layer that leverages ground segmentation to create more reliable occupancy estimates for navigation in

> challenging outdoor environments.

Reviewer: ```suggestion
challenging outdoor environments. While it can be used with any ground segmentation algorithm, we recommend using GSeg3D for which this plugin was originally developed to integrate with.
```

Plug your work, but also make clear that this can be used outside as well

- Very low objects are treated as terrain variation
- Only objects within the robot’s collision range are considered blocking

At times, the terrain is such that no local ground height can be reliably determined. In this case, the
layer can be configured to use neighboring cells to estimate local ground height (see the ``ground_neighbor_search_cells`` parameter), or treat all such obstacles
without ground below them as blocking. For example, if the robot is navigating through a tunnel and the ground
segmentation fails to detect any ground points, then as a backup plan, a ``maximum_height_filter`` (see `Tuning Guide`_) can be applied to incoming obstacle points.
This allows the robot to navigate through tunnels and under bridges without being blocked by misclassified ground points.

3. Temporal Stability Through Evidence Decay
--------------------------------------------

Evidence is decayed over time to allow the costmap to adapt to changing environments. Cells transition gradually between free and
occupied states as evidence builds or fades. The rate of
decay can be tuned separately for ground and obstacle evidence, creating temporal hysteresis, which allows for more stable and

> responsive terrain adaptation (faster ground decay) while maintaining stable obstacle marking (slower obstacle decay).

Reviewer: I won't say no, but I think you should remove this. The bullet points above make this point. I'd insead just add a sentence "if you'd like to learn more, check out these resources" and point to your github and a link to your paper.

This is a bit too much information in my opinion, but would be good for your readme docs if you didn't have it already

Nav2 Integration
================

> 

Reviewer: Then I think this section title can disappear, since we're now configuring your Ground Consistency Layer, which is the name of the prior section. It flows well as a single seciton

> Configure Nav2 to use the ground consistency layer in its local costmap. It is recommended to use the layer in the local costmap

Reviewer: ```suggestion
We will now configure Nav2 to use the ground consistency layer in its local costmap. It is recommended to use the layer in the local costmap
```

Here is an example configuration for the local costmap:

.. code-block:: yaml

   local_costmap:
     local_costmap:
       ros__parameters:
         # ... other costmap settings ...
         plugins: ["ground_consistency", "inflation_layer"]

         ground_consistency:
           plugin: "nav2_ground_consistency_costmap_plugin::GroundConsistencyLayer"
           ground_points_topic: /ground_segmentation/ground_points
           nonground_points_topic: /ground_segmentation/obstacle_points
           tf_timeout: 0.1
           min_clearance: 0.1
           robot_height: 0.92
           maximum_height_filter: 2.0
           ground_inc: 1.0
           nonground_inc: 1.5
           nonground_decay: 0.93
           ground_decay: 0.80
           max_score: 5000.0
           nonground_occ_thresh: 6.0
           nonground_prob_thresh: 0.75
           enable_kpi_logging: false
           discretize_costs: true
           max_data_range: 50.0
           ground_neighbor_search_cells: 0

         inflation_layer:
           plugin: "nav2_costmap_2d::InflationLayer"
           cost_scaling_factor: 3.0
           inflation_radius: 0.7

Parameters Reference

> --------------------

Reviewer: This table does not need to be exhaustive, that's what your docs are for which you can reference at the end in resources.  Basically, remove this table, remove the "Tuning Guide" title breaking it  into a separate section, but keep the Tuning Guide's content. You do a good job there. No reason to duplicate.

     - Set ``maximum_height_filter`` to the maximum obstacle height you want the layer to evaluate; objects above this are treated as overhead/non-blocking
   * - Forest/Cluttered Navigation
     - Branches, leaves, and small debris create false obstacles
     - Only actual blocking-height obstacles are marked
     - Adjust ``min_clearance`` to filter out debris below robot height
   * - Temporal Stability in Dynamic Environments
     - Sensor noise creates flickering obstacles
     - Evidence accumulation and decay smooth the costmap over time
     - Increase decay values to smooth out noise, decrease to respond faster to changes
   * - Varied Terrain with Elevation Changes
     - Traditional costmaps use fixed height thresholds in the global frame, failing on slopes and inclines where ground elevation varies significantly
     - Ground Consistency uses terrain-relative heights, adapting to local ground level rather than absolute global thresholds
     - Set appropriate ``min_clearance`` and ``maximum_height_filter`` values. Ensure ``robot_height`` is accurately set—only obstacles taller than ``robot_height`` relative to ground are considered blocking

> 

Reviewer: Right here is where the "Demo" section could be showing videos / images. I know you have many, here's where you should show them off!
